projects/update_success_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,344 @@
+
+
+```python
+update_dupes = deduped_subset.map(lambda p: ((p.get("clientId"),
+                                              p.get("payload/previousChannel"),
+                                              p.get("payload/previousVersion"),
+                                              p.get("payload/previousBuildId")), 1)).countByKey()
+
+print("Percentage of pings related to the same update (for the same client):\t{:.3f}%"\
+      .format(pct(sum([v for v in update_dupes.values() if v > 1]), deduped_count)))
+```
+    Percentage of pings related to the same update (for the same client):	1.318%
+
+
+We're receiving `update` pings with different `documentId` related to the same initial build, for a few clients. One possible reason for this could be users having multiple copies of Firefox installed on their machine. Let's see if that's the case.
+
+
+```python
+clientIds_sending_dupes = [k[0] for k, v in update_dupes.iteritems() if v > 1]
+
+def check_same_original_build(ping_list):
+    # Build a "unique" identifier for the build by
+    # concatenating the buildId, channel and version.
+    unique_build_ids = [
+        "{}{}{}".format(p.get("application/buildId"), p.get("application/channel"), p.get("application/version"))\
+        for p in ping_list[1]
+    ]
+    
+    # Remove the duplicates and return True if all the pings came
+    # from the same build.
+    return len(set(unique_build_ids)) < 2
+    
+# Count how many duplicates are updating to the same builds and how many are
+# updating to the same builds.
+original_builds_same =\
+    deduped_subset.filter(lambda p: p.get("clientId") in clientIds_sending_dupes)\
+                  .map(lambda p: ((p.get("clientId"),
+                                   p.get("payload/previousChannel"),
+                                   p.get("payload/previousVersion"),
+                                   p.get("payload/previousBuildId")), [p]))\
+                  .reduceByKey(lambda a, b: a + b)\
+                  .filter(lambda p: len(p[1]) > 1)\
+                  .map(check_same_original_build).countByValue()
+                    
+print("Updated builds are identical:\t{:.3f}%"\
+      .format(pct(original_builds_same.get(True), sum(original_builds_same.values()))))
+print("Updated builds are different:\t{:.3f}%"\
+      .format(pct(original_builds_same.get(False), sum(original_builds_same.values()))))
+```
+    Updated builds are identical:	16.472%
+    Updated builds are different:	83.528%
+
+
+The data shows that the `update` pings with the same previous build information may be coming from the same profile, copied and then used with different versions of Firefox. Depending on when the browser is started with a specific copied profile, the downloaded *update* blob might be different (more recent), thus resulting in an `update` with `reason = success` being sent with the same *previous build* information but with different *current build* information.
```

Comment:
Be a little more clear that 83.528% of the 1.318% dupes are updating from different builds. The 0.22% of all update pings that are the same client updating _to_ the same build _from_ the same build are, at present, unexplained. (but in small enough quantities we can ignore for the moment.)

Reply:
Oh, yes, this rephrasing is much better! Thanks, I've updated it!